Modules/posixmodule.c

```diff
@@ -7641,8 +7641,7 @@ os_getgroups_impl(PyObject *module)
     if (n < 0) {
         return posix_error();
     } else {
-        n++; // Avoid malloc(0)
-        grouplist = PyMem_New(gid_t, n+1);
+        grouplist = PyMem_New(gid_t, n);
```

Comment:
n+1 can be removed.

But I think `n++` should be kept to avoid `n = getgropus(0, grouplist)`.
When n=0, getgroups returns number of groups without changing grouplist.
Overallocating one entry is not so bad.

Reply:
@methane It's easy to fix by surrounding:

```c
n = getgroups(n, grouplist);
if (n == -1) {
    PyMem_Free(grouplist);
    return posix_error();
}
```

with: `if (n > 0) { ... }`.

Reply:
I modified my PR to not call getgroups() if n=0. I also cleaned up the code to make it easier to follow.

Reply:
@arhadthedev: See my updated PR. I rewrote the code more it more "flat", easier to follow for the most common path.